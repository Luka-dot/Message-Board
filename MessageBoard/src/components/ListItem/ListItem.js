import React from 'react';
import checked from '../../media/checked.svg';
import unchecked from '../../media/unchecked.svg';

export const ListItem = (props) => {

  const { id, index, isComplete, text, handleItemChange, handleItemDelete, toggleComplete } = props;
  let listType = isComplete ? 'ItemComplete' : 'ItemIncomplete';

  return (
    <li key={index} className={listType} >
      <button className="CheckBox" onClick={() => toggleComplete(id)}>
        <img src={isComplete ? checked : unchecked} alt="Check Box" />
      </button>
      <input type="text"
             placeholder="List Item"
             value={text}
             autoComplete="off"
             name="ListItem"
             className="list-item-input"
             onChange={(e) => handleItemChange(e, id)}
             disabled={isComplete ? 'disabled' : null}
             id={id}
        />
        <button className="delete-button" onClick={(e) => handleItemDelete(e, id)}>
          <i className="fas fa-times"></i>
        </button>
    </li>
  )

}
