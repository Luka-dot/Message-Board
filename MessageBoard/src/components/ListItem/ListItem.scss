ul.ListItems {
  margin: 0;
  padding: 0;

  button.CheckBox {
    transition: all .5s ease;

    &:hover {
      opacity: .7;
    }
  }

  li.ItemIncomplete {
    position: relative;
    transition: all .3 ease;

    &:last-child {

      &:after {
        content: '';
        position: absolute;
        height: 1px;
        width: 92%;
        bottom: 0px;
        border-bottom: 1px solid rgba(0, 0, 0, 0.15);
        left: 0;
        right: 0;
        margin-left: auto;
        margin-right: auto;
      }

      .CheckBox {
        img { display: none; }
        position: relative;
        pointer-events: none;
        background: url('./media/plus.svg') no-repeat;
        left: 13.5px;
        top: 4px;
        opacity: .5;
      }

      input.list-item-input {
        width: 90%;
        &::placeholder {
          color: $textColor;
        }
      }

      .delete-button {
        display: none
      }
    }

    &:not(:last-child) {

      .delete-button {
        opacity: 0;
        pointer-events: none;
        transition: all .3 ease;
      }

      &:hover {
        .delete-button {
          opacity: .6;
          pointer-events: all;
          &:hover {
            opacity: .9;
          }
        }
      }

      &:focus-within {

        .delete-button {
          opacity: .6;
          pointer-events: all;
        }

        &:before {
          content: '';
          position: absolute;
          height: 1px;
          width:100%;
          top: 0px;
          border-top: 1px solid rgba(0, 0, 0, 0.3);
        }
        &:after {
          content: '';
          position: absolute;
          height: 1px;
          width:100%;
          bottom: 0px;
          border-bottom: 1px solid rgba(0, 0, 0, 0.3);
        }

      }

      input.list-item-input {
        text-decoration: none;

        &::placeholder {
          color: transparent;
        }
      }

      .CheckBox:hover + .list-item-input {
        text-decoration: line-through;
      }

    }

  }

  li.ItemComplete {

    input.list-item-input {
      text-decoration: line-through;
    }

    .CheckBox:hover + .list-item-input {
      text-decoration: none;
    }
  }

  li {
    list-style-type: none;
    display: flex;
    justify-content: space-between;
    padding: 3px 0;

    .list-item-input {
      font-family: $font-Roboto;
      font-size: .9rem;
      font-weight: 400;
      color: $textColor;
      line-height: 1.5rem;
    }

    label {
      width: 4%;
    }

    input[type=text] {
      outline: none;
      border:none;
      width: 80%;
      background: inherit;
    }

    button {
      background: none;
      outline: none;
      width: 10%;
      border: none;
      cursor: pointer;
      line-height: 1;
      position: relative;
      bottom: -2px;
    }

    img {
      height: 18px;
      opacity: .54;
    }
  }
}
