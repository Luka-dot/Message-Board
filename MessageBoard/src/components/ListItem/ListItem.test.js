import React from 'react';
import { shallow } from 'enzyme';
import { ListItem } from './ListItem';

describe('ListItem', () => {
  let wrapper;

  let mockItem = {
    id: "1234",
    isComplete: false,
    text: "Add basic styling"
  }
  const mockIndex = 0
  const mockId = "123232"
  let mockisComplete = false
  const mockHandleItemChange = jest.fn()
  const mockHandleItemDelete = jest.fn()
  const mockToggleComplete = jest.fn()

  beforeEach(() => {
    wrapper = shallow(
      <ListItem
        item={mockItem}
        id={mockId}
        isComplete={mockisComplete}
        index={mockIndex}
        toggleComplete={mockToggleComplete}
        handleItemChange={mockHandleItemChange}
        handleItemDelete={mockHandleItemDelete}
      />)
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should call toggleComplete when the checkbox is checked', () => {
    let input = wrapper.find('.CheckBox')
    input.simulate('click')
    expect(mockToggleComplete).toHaveBeenCalledWith(mockId)
  })

  it('should call handleItemChange when list item input is changed', () => {
    let input = wrapper.find('.list-item-input')
    input.simulate('change')
    expect(mockHandleItemChange).toHaveBeenCalled()
  })

  it('should call handleItemDelete when the delete button is clicked', () => {
    let button = wrapper.find('.delete-button')
    button.simulate('click')
    expect(mockHandleItemDelete).toHaveBeenCalled()
  })

  it('should match the snapshot when an item isCompleted', () => {
    wrapper = shallow(<ListItem isComplete={true} />)
    expect(wrapper).toMatchSnapshot()
  })
})
