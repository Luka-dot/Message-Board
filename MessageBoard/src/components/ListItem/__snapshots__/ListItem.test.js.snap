// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem should match the snapshot 1`] = `
<li
  className="ItemIncomplete"
  key="0"
>
  <button
    className="CheckBox"
    onClick={[Function]}
  >
    <img
      alt="Check Box"
      src="unchecked.svg"
    />
  </button>
  <input
    autoComplete="off"
    className="list-item-input"
    disabled={null}
    id="123232"
    name="ListItem"
    onChange={[Function]}
    placeholder="List Item"
    type="text"
  />
  <button
    className="delete-button"
    onClick={[Function]}
  >
    <i
      className="fas fa-times"
    />
  </button>
</li>
`;

exports[`ListItem should match the snapshot when an item isCompleted 1`] = `
<li
  className="ItemComplete"
>
  <button
    className="CheckBox"
    onClick={[Function]}
  >
    <img
      alt="Check Box"
      src="checked.svg"
    />
  </button>
  <input
    autoComplete="off"
    className="list-item-input"
    disabled="disabled"
    name="ListItem"
    onChange={[Function]}
    placeholder="List Item"
    type="text"
  />
  <button
    className="delete-button"
    onClick={[Function]}
  >
    <i
      className="fas fa-times"
    />
  </button>
</li>
`;
